nexus/src/app/instance.rs

```diff
@@ -774,6 +774,108 @@ impl super::Nexus {
             .map_err(Into::into)
     }
 
+    /// Forcefully stop a running instance, causing its sled-agent to rudely
+    /// terminate its VMM process and unregister the instance.
+    pub(crate) async fn instance_force_terminate(
+        &self,
+        opctx: &OpContext,
+        instance_lookup: &lookup::Instance<'_>,
+    ) -> Result<InstanceAndActiveVmm, InstanceStateChangeError> {
+        let (.., authz_instance) =
+            instance_lookup.lookup_for(authz::Action::Modify).await?;
+
+        let state = self
+            .db_datastore
+            .instance_fetch_with_vmm(opctx, &authz_instance)
+            .await?;
+
+        // Is the instance currently incarnated in a VMM process?
+        let Some(vmm) = state.vmm() else {
+            // If the instance is already un-incarnated, we're all good.
+            debug!(
+                opctx.log,
+                "asked to force terminate an instance that has no VMM, \
+                 doing nothing";
+                "instance_id" => %authz_instance.id(),
+                "instance_state" => ?state.instance.runtime(),
+            );
+            debug_assert_eq!(
+                state.instance.runtime().nexus_state,
+                db::model::InstanceState::NoVmm,
+                "if an instance has no active VMM, it must be in the NoVmm \
+                 state (this is enforced by a DB check constraint)",
+            );
+            return Ok(state);
+        };
+
+        // Ladies and gentlemen, we got him!
+        let propolis_id = PropolisUuid::from_untyped_uuid(vmm.id);
+        let sled_id = SledUuid::from_untyped_uuid(vmm.sled_id);
+        let unregister_result =
+            self.instance_ensure_unregistered(&propolis_id, &sled_id).await;
```

Comment:
Does the start saga clean up properly if this happens between its `ensure_registered` and `ensure_running` steps? I *think* it works out: `sis_ensure_running` will fail; `sis_ensure_registered_undo` will also fail and try to move the VMM to Failed, but this doesn't block the saga from continuing to unwind; then I think we'll end up in SagaUnwound and end up with a VM that can be started again. Does that sound about right? If so, it might be worthwhile to add a comment mentioning this case.

Reply:
I *think* so, but I would like to figure out whether it's possible to test this...

Reply:
Can we do something with error types instead? That is: have `mark_vmm_failed` return its own error enum with "DB query failed" and "update too old" variants, and have `sis_ensure_registered_undo` match on specific error codes from this callee, such that a new failure mode will break the match. WDYT? It'd be nice to have an integration test, too, but I'm similarly having trouble figuring out how to inject a failure into this specific call, since we don't have a CRDB test double that I know of.

Reply:
Hm, actually, the write to CRDB in `mark_vmm_failed` is done by a call to `vmm_update_runtime`, which returns `Ok(false)` if the VMM exists but wasn't updated (e.g. if the generation has advanced). So, we don't even hit the error path (which gets ignored anyway) in `mark_vmm_failed` in that case...

Reply:
Which, now that I look at it, indicates a tiny possible inefficiency in `mark_vmm_failed` --- currently, we try to run an update saga even if the VMM did not move to failed (because it was already failed/destroyed), which means we will probably start spurious update sagas there. I'll clean that up.

Reply:
I think I'm probably fine with having `mark_vmm_failed` stick with the convention of returning `Ok(false)` if the generation has changed, instead of an error variant, since it seems like we generally follow that convention for similar code. On one hand, I do have a sort of ideological preference for an `Ok(_)` to _always_ mean "yes, we actually wrote the desired update to CRDB", but on the other hand, a lot of saga undo actions probably just `?` these functions, and would probably prefer to get an `Ok` in any case that doesn't mean they should unwind. I dunno.

Reply:
`mark_vmm_failed` actually has the type signature of `async fn(...) -> Result<(), Error>` but there are actually no conditions in which it will ever currently return an error. That's cool.

I might just change the type signature to not return `Result`

Reply:
Upon further perusal of `sis_ensure_registered_undo`, I noticed that it was actually handling sled-agent unregister errors in a pretty outdated way: it was treating `vmm_gone` errors as a probable failure, and just ignoring any other errors communicating with the sled-agent, as described in this comment:

https://github.com/oxidecomputer/omicron/blob/0640bb277df110f3e740464bf4bacdf2bd24c897/nexus/src/app/sagas/instance_start.rs#L673-L691

This predates the changes to `vmm_gone`/`instance_unhealthy` semantics from RFD 486/#6455, and isn't quite correct. I've changed it to make the "other error from sled agent" be the case that fails the saga, although we should probably retry communication errors eventually. 

The changes to `mark_vmm_failed` and `sis_ensure_registered_undo` are in a923f2a6a058fc1b42da33c8d1f52da6592b0bd1. Arguably, it's not *really* in scope for this PR at that point; I'd be happy to cherry-pick it out to a separate branch if you think that's worth doing?